src/net.cpp

```diff
@@ -776,7 +791,160 @@ CNetMessage V1TransportDeserializer::GetMessage(const std::chrono::microseconds
+
+    memcpy(&vRecv[m_hdr_pos], msg_bytes.data(), copy_bytes);
+    m_hdr_pos += copy_bytes;
+
+    // if AAD incomplete, exit
+    if (m_hdr_pos < CHACHA20_POLY1305_AEAD_AAD_LEN) {
+        return copy_bytes;
+    }
+
+    // we got the AAD bytes at this point (3 bytes encrypted packet length)
+    m_message_size = m_aead->DecryptLength((const uint8_t*)vRecv.data());
+
+    // reject messages larger than MAX_SIZE
+    if (m_message_size > V2_MAX_PAYLOAD_LENGTH) {
+        return -1;
+    }
+
+    // switch state to reading message data
+    m_in_data = true;
+
+    return copy_bytes;
+}
+int V2TransportDeserializer::readData(Span<const uint8_t> msg_bytes)
+{
+    // Read the message data (command, payload & MAC)
+    const unsigned int remaining = m_message_size + CHACHA20_POLY1305_AEAD_TAG_LEN - m_data_pos;
+    const unsigned int copy_bytes = std::min<unsigned int>(remaining, msg_bytes.size());
+
+    // extend buffer, respect previous copied AAD part
+    if (vRecv.size() < CHACHA20_POLY1305_AEAD_AAD_LEN + m_data_pos + copy_bytes) {
+        // Allocate up to 256 KiB ahead, but never more than the total message size (incl. AAD & TAG).
+        vRecv.resize(CHACHA20_POLY1305_AEAD_AAD_LEN + std::min(m_message_size, m_data_pos + copy_bytes + 256 * 1024) + CHACHA20_POLY1305_AEAD_TAG_LEN);
+    }
+
+    memcpy(&vRecv[CHACHA20_POLY1305_AEAD_AAD_LEN + m_data_pos], msg_bytes.data(), copy_bytes);
+    m_data_pos += copy_bytes;
+
+    return copy_bytes;
+}
+
+
+CNetMessage V2TransportDeserializer::GetMessage(const std::chrono::microseconds time, bool& reject_message, bool& disconnect)
+{
+    // Initialize out parameters
+    reject_message = false;
+    disconnect = false;
+
+    // In v2, vRecv contains the encrypted length (AAD), encrypted payload plus the MAC tag
+    // (3 bytes AAD + 1-13 bytes serialized message command + ? bytes message payload + 16 byte MAC tag)
+    assert(Complete());
+
+    std::string command_name;
+
+    // defensive decryption (MAC check, decryption, command deserialization)
+    // we'll always return a CNetMessage (even if decryption fails)
+    if (m_aead->Crypt((unsigned char*)vRecv.data(), vRecv.size(), (const uint8_t*)vRecv.data(), vRecv.size(), false)) {
+        // MAC check was successful
+        // We can remove packet length (stored in m_message_size) and MAC tag (verified)
+        assert(vRecv.size() >= CHACHA20_POLY1305_AEAD_AAD_LEN + CHACHA20_POLY1305_AEAD_TAG_LEN);
+
+        // CDataStream::erase at the begin will just increase the read pos
+        vRecv.erase(vRecv.begin(), vRecv.begin() + CHACHA20_POLY1305_AEAD_AAD_LEN);
+        vRecv.erase(vRecv.end() - CHACHA20_POLY1305_AEAD_TAG_LEN, vRecv.end());
+
+        uint8_t size_or_shortid = 0;
+        try {
+            vRecv >> size_or_shortid;
+        } catch (const std::ios_base::failure&) {
+            LogPrint(BCLog::NET, "Invalid message type, peer=%d\n", m_node_id);
+            reject_message = true;
```

Comment:
So in this case, `size_or_shortid` will stay at 0, and it can simply continue below?

Reply:
Yes. It makes some of the invariants about `msg` below hold and that makes fuzz testing easier. The client of this function is expected to reject the message in this case.

Reply:
Thanks, clear.